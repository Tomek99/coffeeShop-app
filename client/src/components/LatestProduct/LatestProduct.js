import React, { useContext, useEffect, useState } from "react";
import { Context } from "../../Contexts/Context";
import { BsEye, BsCart, BsHeart } from "react-icons/bs";
import { useNavigate } from "react-router-dom";
import styles from "./LatestProduct.module.scss";
import PropTypes from "prop-types";
import RatingsStars from "../RatingStars/RatingStars";
import BtnsProduct from "../Buttons/BtnsProduct/BtnsProduct";

function LatestProduct({ item }) {
  const { _id, imageUrl, name, price, oldPrice, productRatings, intensity } =
    item;
  const { addItem, addWishItem, saveViewedProduct, wishList } =
    useContext(Context);

  const viwedProduct = {
    productId: _id,
    productUrl: imageUrl,
    productName: name,
  };

  const navigate = useNavigate();
  function navigatePage() {
    navigate(`/products/${_id}`);
    saveViewedProduct(viwedProduct);
  }

  const productRatingsAverage =
    productRatings && productRatings.length > 0
      ? productRatings.reduce((sum, current) => sum + current, 0) /
        productRatings.length
      : 0;
  return (
    <div className={styles.LatestProduct}>
      <BtnsProduct
        item={item}
        wishList={wishList}
        navigatePage={navigatePage}
        addItem={addItem}
        addWishItem={addWishItem}
      />
      <div className={styles.contentWrapper}>
        <img src={imageUrl} alt={name} />
        <h3 className={styles.productName} onClick={navigatePage}>
          {name}
        </h3>
        <p className={styles.intensity}>
          Intensity <span>{intensity}</span>
        </p>
        <div className={styles.divRatingsStars}>
          <RatingsStars rate={productRatingsAverage} size="large" />
          <span>({productRatings.length})</span>
        </div>
      </div>
      <p className={styles.price}>
        <span>${price} </span>
        <span className={styles.oldPrice}>
          {Boolean(oldPrice) ? `$${oldPrice}` : null}
        </span>
      </p>
    </div>
  );
}

LatestProduct.propTypes = {
  id: PropTypes.number,
  imageUrl: PropTypes.string,
  name: PropTypes.string,
  newPrice: PropTypes.number,
  oldPrice: PropTypes.number,
  addItem: PropTypes.func,
  quantity: PropTypes.number,
  rate: PropTypes.number,
};
export default LatestProduct;
