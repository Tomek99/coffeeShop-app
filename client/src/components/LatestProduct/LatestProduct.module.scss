.LatestProduct {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  width: 300px;
  align-items: center;
  text-align: center;
  border-radius: 8px;
  padding: 30px 0;
  transition: 0.3s linear;

  &:hover {
    box-shadow: var(--shadow-thin);
    background-color: var(--light-greay);
  }

  .contentWrapper {
    img {
      width: 100%;
      max-width: 120px;
      height: 210px;
      object-fit: cover;
    }

    .divRatingsStars {
      display: flex;
      justify-content: center;
      align-items: center;
    }

    .productName {
      display: flex;
      justify-content: center;
      align-items: center;
      margin: 0 10px;
      height: 40px;
      font-size: 1.7rem;
      transition: 0.2s linear;
      position: relative;
      cursor: pointer;

      // &::before {
      //   content: "";
      //   position: absolute;
      //   bottom: -1.5px;
      //   left: 50%;
      //   right: 50%;
      //   width: 0;
      //   height: 2px;
      //   background-color: var(--text-color);
      //   transition: all 0.3s ease;
      // }
      &:hover {
        text-decoration: underline;
      }
      // &:hover::before {
      //   width: 100%;
      //   left: 0;
      //   right: 0;
      // }
    }

    .intensity {
      display: flex;
      justify-content: center;
      align-items: center;
      gap: 5px;
      font-size: 1.3rem;
      font-weight: 300;
      text-transform: uppercase;
      color: var(--text-color);
      margin-bottom: 5px;

      span {
        display: flex;
        justify-content: center;
        align-items: center;
        height: 25px;
        width: 25px;
        box-shadow: var(--shadow-thin);
        border-radius: 50%;
        font-weight: bold;
      }
    }
  }

  .price {
    font-size: 2.2rem;
    color: var(--text-color);

    .oldPrice {
      font-size: 1.5rem;
      text-decoration: line-through;
      color: var(--text-color);
      opacity: 0.7;
    }
  }
}
