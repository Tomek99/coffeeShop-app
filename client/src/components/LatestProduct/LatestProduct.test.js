import { render, screen } from "@testing-library/react";
import LatestProduct from "./LatestProduct";
import { Context } from "../../Contexts/Context";
import { MemoryRouter } from "react-router-dom";

window.scrollTo = jest.fn();

test("should render LatestProduct component", () => {
  const item = {
    _id: "3gijh3g0h23",
    imageUrl: "",
    name: "test",
    price: 13.5,
    oldPrice: null,
    rate: 4,
    intensity: 7,
    productRatings: [2, 3],
  };

  const wishList = [];
  const { container } = render(
    <Context.Provider value={{ wishList }}>
      <MemoryRouter>
        <LatestProduct item={item} />
      </MemoryRouter>
    </Context.Provider>
  );

  expect(container).toMatchSnapshot();
});
