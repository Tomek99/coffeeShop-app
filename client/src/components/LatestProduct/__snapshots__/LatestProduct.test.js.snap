// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render LatestProduct component 1`] = `
<div>
  <div
    class="LatestProduct"
  >
    <div
      class="BtnsProduct"
    >
      <button>
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 8s-3-5.5-8-5.5S0 8 0 8s3 5.5 8 5.5S16 8 16 8zM1.173 8a13.133 13.133 0 0 1 1.66-2.043C4.12 4.668 5.88 3.5 8 3.5c2.12 0 3.879 1.168 5.168 2.457A13.133 13.133 0 0 1 14.828 8c-.058.087-.122.183-.195.288-.335.48-.83 1.12-1.465 1.755C11.879 11.332 10.119 12.5 8 12.5c-2.12 0-3.879-1.168-5.168-2.457A13.134 13.134 0 0 1 1.172 8z"
          />
          <path
            d="M8 5.5a2.5 2.5 0 1 0 0 5 2.5 2.5 0 0 0 0-5zM4.5 8a3.5 3.5 0 1 1 7 0 3.5 3.5 0 0 1-7 0z"
          />
        </svg>
      </button>
      <button>
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 1.5A.5.5 0 0 1 .5 1H2a.5.5 0 0 1 .485.379L2.89 3H14.5a.5.5 0 0 1 .491.592l-1.5 8A.5.5 0 0 1 13 12H4a.5.5 0 0 1-.491-.408L2.01 3.607 1.61 2H.5a.5.5 0 0 1-.5-.5zM3.102 4l1.313 7h8.17l1.313-7H3.102zM5 12a2 2 0 1 0 0 4 2 2 0 0 0 0-4zm7 0a2 2 0 1 0 0 4 2 2 0 0 0 0-4zm-7 1a1 1 0 1 1 0 2 1 1 0 0 1 0-2zm7 0a1 1 0 1 1 0 2 1 1 0 0 1 0-2z"
          />
        </svg>
      </button>
      <button>
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m8 2.748-.717-.737C5.6.281 2.514.878 1.4 3.053c-.523 1.023-.641 2.5.314 4.385.92 1.815 2.834 3.989 6.286 6.357 3.452-2.368 5.365-4.542 6.286-6.357.955-1.886.838-3.362.314-4.385C13.486.878 10.4.28 8.717 2.01L8 2.748zM8 15C-7.333 4.868 3.279-3.04 7.824 1.143c.06.055.119.112.176.171a3.12 3.12 0 0 1 .176-.17C12.72-3.042 23.333 4.867 8 15z"
          />
        </svg>
      </button>
    </div>
    <div
      class="contentWrapper"
    >
      <img
        alt="test"
        src=""
      />
      <h3
        class="productName"
      >
        test
      </h3>
      <p
        class="intensity"
      >
        Intensity 
        <span>
          7
        </span>
      </p>
      <div
        class="divRatingsStars"
      >
        <span
          aria-label="3 Stars"
          class="MuiRating-root MuiRating-sizeLarge MuiRating-readyOnly RatingsStars css-1x1lh1c-MuiRating-root"
          role="img"
        >
          <span>
            <span
              class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
          </span>
          <span>
            <span
              class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
          </span>
          <span>
            <span
              class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
          </span>
          <span>
            <span
              class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarBorderIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                />
              </svg>
            </span>
          </span>
          <span>
            <span
              class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarBorderIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                />
              </svg>
            </span>
          </span>
        </span>
        <span>
          (
          2
          )
        </span>
      </div>
    </div>
    <p
      class="price"
    >
      <span>
        $
        13.5
         
      </span>
      <span
        class="oldPrice"
      />
    </p>
  </div>
</div>
`;
